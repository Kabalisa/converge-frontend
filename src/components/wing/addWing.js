/* eslint-disable react/prop-types */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import toastr from 'toastr';
import MrmModal from '../commons/Modal';
import { Input, SelectInput as Select } from '../commons';
import '../../assets/styles/addoffice.scss';
import ActionButtons from '../commons/ActionButtons';
import ADD_WING_MUTATION from '../../graphql/mutations/wings';
import GET_FLOORS_QUERY from '../../graphql/queries/Floors';
import notification from '../../utils/notification';

/**
 * Add Wing Component
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
export class AddWing extends Component {
  static propTypes = {
    addWing: PropTypes.func.isRequired,
  };

  state = {
    name: '',
    floorId: '',
    closeModal: false,
    isLoading: false,
  };

  /**
   * Ensures that the modal for adding wing closes
   *
   * @returns {void}
   */
  handleCloseModal = () => {
    this.setState({ closeModal: true });
  };

  /**
   * Ensures that the state is updated basing
   * on the changes in the input fields
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleInputChange = (event) => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  /**
   * It updates the state value of closeModal
   * to false whenever the modal closes
   *
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  /**
   * Handles the adding of Wings
   *
   * @returns {void}
   */
  handleAddWing = (event) => {
    event.preventDefault();
    const { name, floorId } = this.state;
    if (!name) {
      notification(toastr, 'error', 'Wing name is required')();
    } else {
      this.toggleLoading();
      this.props
        .addWing({
          variables: {
            name,
            floorId,
          },
        })
        .then((wing) => {
          this.toggleLoading();
          this.handleCloseModal();
          notification(
            toastr,
            'success',
            `${wing.data.addWing.wing.name} wing has been added successfully`,
          )();
        })
        .catch((err) => {
          this.toggleLoading();
          this.handleCloseModal();
          notification(toastr, 'error', err.graphQLErrors[0].message)();
        });
    }
  };

  /**
   * 1. change isLoading state to it's opposite value
   * i.e true to false or vise verser
   *
   * @returns {void}
   */
  toggleLoading = () => {
    this.setState({
      isLoading: !this.state.isLoading,
    });
  }

  render() {
    const {
      name, floorId, closeModal, isLoading,
    } = this.state;
    const { allFloors } = this.props.allFloors;
    const { handleAddWing, handleInputChange, handleCloseModal } = this;

    return (
      <MrmModal
        title="ADD WING"
        buttonText="Add Wing"
        closeModal={closeModal}
        handleCloseRequest={this.handleModalStateChange}
        className="add-office-modal"
        modalButton="add-button"
      >
        <form className="modal-form">
          <Input
            labelName="Wing Name"
            name="name"
            value={name}
            placeholder="Enter wing name"
            id="wingName"
            onChange={handleInputChange}
          />
          <Select
            labelText="Select floor"
            name="floorId"
            id="floorId"
            value={floorId}
            onChange={handleInputChange}
            wrapperClassName="input-wrapper"
            placeholder="Select floor"
            options={allFloors}
          />
          <ActionButtons
            withCancel
            isLoading={isLoading}
            onClickCancel={handleCloseModal}
            actionButtonText="ADD WING"
            onClickSubmit={handleAddWing}
          />
        </form>
      </MrmModal>
    );
  }
}

export default compose(
  graphql(GET_FLOORS_QUERY, { name: 'allFloors' }),
  graphql(ADD_WING_MUTATION, { name: 'addWing' }),
)(AddWing);
