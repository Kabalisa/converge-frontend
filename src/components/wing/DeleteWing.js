import React from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import toastr from 'toastr';
import MrmModal from '../commons/Modal';
import { DELETE_WING_MUTATION } from '../../graphql/mutations/wings';
import { GET_ALL_WINGS } from '../../graphql/queries/wings';
import notification from '../../utils/notification';
import ActionButtons from '../commons/ActionButtons';
import '../../assets/styles/deleteModal.scss';

/**
 * Delete Wing Component
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
export class DeleteWing extends React.Component {
  state = {
    closeModal: false,
    isDeleting: false,
  };

  /**
   * Ensures that the modal for deleting wing closes
   *
   * @returns {void}
   */
  handleCloseModal = () => {
    this.setState({ closeModal: true });
  };

  /**
   * Handles the state changes for the deleting wing modal
   *
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  /**
   * Handles deleting wing
   *
   * @returns {void}
   */
  handleDeleteWing = () => {
    const { wingId, deleteWing } = this.props;
    this.toggleLoading();
    deleteWing({
      variables: {
        wingId,
      },
      refetchQueries: [{ query: GET_ALL_WINGS }],
    })
      .then((wing) => {
        this.toggleLoading();
        this.handleCloseModal();
        const { name } = wing.data.deleteWing.wing;
        notification(toastr, 'success', `${name} is deleted successfully`)();
      })
      .catch((err) => {
        this.toggleLoading();
        this.handleCloseModal();
        notification(toastr, 'error', err.graphQLErrors[0].message)();
      });
  };

  /**
   * 1. change isLoading state to it's opposite value
   * i.e true to false or vise verser
   *
   * @returns {void}
   */
  toggleLoading = () => {
    this.setState({
      isDeleting: !this.state.isDeleting,
    });
  }

  render() {
    const { isDeleting } = this.state;
    return (
      <MrmModal
        handleCloseRequest={this.handleModalStateChange}
        className="delete-modal"
        buttonText="Delete"
        closeModal={this.state.closeModal}
        title="DELETE WING"
      >
        <div className="delete-modal-content">
          <p id="confirm-msg">
            Are you sure you want to delete the {`"${this.props.wingName}"`}{' '}
            wing? This cannot be undone & all data will be lost
          </p>
          <ActionButtons
            withCancel
            onClickCancel={this.handleCloseModal}
            isLoading={isDeleting}
            actionButtonText="DELETE WING"
            onClickSubmit={this.handleDeleteWing}
          />
        </div>
      </MrmModal>
    );
  }
}

DeleteWing.propTypes = {
  deleteWing: PropTypes.func.isRequired,
  wingName: PropTypes.string.isRequired,
  wingId: PropTypes.string.isRequired,
};

export default compose(
  graphql(DELETE_WING_MUTATION, { name: 'deleteWing' }),
  graphql(GET_ALL_WINGS, { name: 'allWings' }),
)(DeleteWing);
