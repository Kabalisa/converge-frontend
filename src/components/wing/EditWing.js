import React from 'react';
import PropTypes from 'prop-types';
import toastr from 'toastr';
import { graphql } from 'react-apollo';
import MrmModal from '../commons/Modal';
import ActionButtons from '../commons/ActionButtons';
import { Input } from '../commons';
import '../../assets/styles/editoffice.scss';
import notification from '../../utils/notification';
import { GET_ALL_WINGS } from '../../graphql/queries/wings';
import { EDIT_WING_MUTATION } from '../../graphql/mutations/wings';

/**
 * Edit Wing Component
 *
 * @extends React.Component
 *
 * @returns {JSX}
 */
export class EditWing extends React.Component {
  static propTypes = {
    wingId: PropTypes.string.isRequired,
    wingName: PropTypes.string.isRequired,
    editWing: PropTypes.func.isRequired,
  };

  state = {
    wingName: this.props.wingName,
    wingId: this.props.wingId,
    closeModal: false,
    isLoading: false,
  };

  /**
   * Ensures that the state is updated basing
   * on the changes in the input fields
   *
   * @param {Object} target
   *
   * @returns {void}
   */
  handleInputChange = ({ target: { name, value } }) => {
    this.setState({ [name]: value });
  };

  /**
   * It closes a modal
   *
   * @returns {void}
   */
  handleCloseModal = () => {
    this.setState({ closeModal: true });
  };

  /**
   * It changes the state of the modal
   *
   * @returns {void}
   */
  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  /**
   * Handles editing wing
   *
   * @param {object} event
   *
   * @returns {void}
   */
  handleEditWing = (e) => {
    const { wingId, wingName } = this.state;
    const { editWing } = this.props;
    e.preventDefault();
    this.toggleLoading();
    editWing({
      variables: {
        wingId,
        name: wingName,
      },
    }).then(() => {
      this.toggleLoading();
      this.handleCloseModal();
      notification(toastr, 'success', `${wingName} office has been updated successfully`)();
    }).catch((err) => {
      this.toggleLoading();
      this.handleCloseModal();
      this.setState({
        wingName: this.props.wingName,
      });
      notification(toastr, 'error', err.graphQLErrors[0].message)();
    });
  }

  /**
 * 1. change isLoading state to it's opposite value
 * i.e true to false or vise verser
 *
 * @returns {void}
 */
  toggleLoading = () => {
    this.setState({
      isLoading: !this.state.isLoading,
    });
  }
  render() {
    const { closeModal, wingName, isLoading } = this.state;
    return (
      <MrmModal
        closeModal={closeModal}
        handleCloseRequest={this.handleModalStateChange}
        title="EDIT WING"
        buttonText="Edit"
        className="edit-office-modal"
        modalButtonClassName="edit-button"
      >
        <form className="modal-form">
          <Input
            name="wingName"
            labelName="Wing Name"
            onChange={this.handleInputChange}
            inputClass="input1"
            value={wingName}
            labelClass="label1"
            id="wingName"
            required
          />
          <ActionButtons
            withCancel
            onClickCancel={this.handleCloseModal}
            actionButtonText="SAVE CHANGES"
            isLoading={isLoading}
            onClickSubmit={this.handleEditWing}
          />
        </form>
      </MrmModal>
    );
  }
}

export default graphql(EDIT_WING_MUTATION, {
  name: 'editWing',
  options: { refetchQueries: [{ query: GET_ALL_WINGS }] },
})(EditWing);
