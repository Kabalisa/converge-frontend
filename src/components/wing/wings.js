import React from 'react';
import PropTypes from 'prop-types';
import DeleteWing from './DeleteWing'; //eslint-disable-line
import EditWing from './EditWing'; //eslint-disable-line

/**
 * Wing component
 *
 * @param {object} wing
 *
 * @returns {JSX}
 */
export const Wing = ({ wing }) => (
  <div className="table__row">
    <span>{wing.name}</span>
    <span>{wing.floor && wing.floor.block.name}</span>
    <span>{wing.floor && wing.floor.name}</span>
    <span>
      <EditWing id="edit-modal" wingName={wing.name} wingId={wing.id} wingBlock={wing.block} />
      &nbsp;
      <DeleteWing wingName={wing.name} id="delete-modal" wingId={wing.id} />
    </span>
  </div>
);
Wing.propTypes = {
  wing: PropTypes.shape({
    name: PropTypes.string.isRequired,
  }).isRequired,
};
export default Wing;
