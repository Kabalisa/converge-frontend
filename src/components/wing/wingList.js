import React from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import Wing from "./wings"; // eslint-disable-line
import AddWing from './addWing'; // eslint-disable-line
import '../../assets/styles/officelist.scss';
import ColGroup from '../helpers/ColGroup';
import TableHead from '../helpers/TableHead';
import MenuTitle from '../commons/MenuTitle';
import Spinner from '../commons/Spinner';
import { decodeTokenAndGetUserData } from '../../utils/Cookie';
import { GET_USER_QUERY } from '../../graphql/queries/People';
import { GET_ALL_WINGS } from '../../graphql/queries/wings';
import Errors from '../commons/Errors';

/**
 * Get user data from token
*/
const { UserInfo: userData } = decodeTokenAndGetUserData() || {};

/**
 * Wing list component
 *
 * @param {array} props
 *
 * @returns {JSX}
 */
export const WingList = (props) => {
  const { allWings, loading, error } = props.allWings;
  if (error) return <Errors message="Data cannot be returned at the moment" />;
  return (
    /* istanbul ignore next */
    loading ? (
      <Spinner />
    ) : (
      <div>
        <div className="settings-offices">
          <div className="settings-offices-control">
            <MenuTitle title="Wings" />
            <AddWing userLocation={props.data.loading ? '' : props.data.user.location} />
          </div>
          <div className="settings-offices-list">
            <table>
              <ColGroup />
              <TableHead titles={['Wing', 'Block', 'Floor', 'Action']} />
              <tbody>
                {allWings &&
                  allWings.map(wing => (
                    <Wing wing={wing} key={wing.name} wingId={wing.id} />
                  ))}
              </tbody>
            </table>
          </div>
        </div>
      </div>
    )
  );
};

WingList.propTypes = {
  allWings: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
  }),
  data: PropTypes.shape({
    loading: PropTypes.bool,
    user: PropTypes.object,
  }),
};

WingList.defaultProps = {
  allWings: {
    id: 1,
    name: '',
  },
  data: {
    loading: false,
  },
};

/* istanbul ignore next */
const options = () => ({
  variables: {
    email: userData.email,
  },
});

export default compose(
  graphql(GET_ALL_WINGS, { name: 'allWings' }),
  graphql(GET_USER_QUERY, {
    options,
  }),
)(WingList);
