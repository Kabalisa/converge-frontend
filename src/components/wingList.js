import React from 'react';
import PropTypes from 'prop-types';
import { graphql } from 'react-apollo';
import Wing from './wings'; // eslint-disable-line
import AddWing from './addWing'; // eslint-disable-line
import '../assets/styles/officelist.scss';
import ColGroup from './helpers/ColGroup';
import TableHead from './helpers/TableHead';
import { GET_ALL_WINGS } from '../graphql/queries/wings';
import MenuTitle from './MenuTitle';
import Spinner from './commons/Spinner';

export const WingList = (props) => {
  const { allWings, loading } = props.allWings;
  return (
    /* istanbul ignore next */
    loading ? <Spinner /> :
    <div>
      <div className="settings-offices">
        <div className="settings-offices-control">
          <MenuTitle title="Wings" />
          <AddWing />
        </div>
        <div className="settings-offices-list">
          <table>
            <ColGroup />
            <TableHead titles={['Wing', 'Block', 'Floor', 'Action']} />
            <tbody>
              {allWings && allWings.map(wing => (
                <Wing
                  wing={wing}
                  key={wing.name}
                  wingId={wing.id}
                />
              ))}
            </tbody>
          </table>
        </div>
      </div>
    </div>
  );
};

WingList.propTypes = {
  allWings: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
  }),
};

WingList.defaultProps = {
  allWings: {
    id: 1,
    name: '',
  },
};

export default graphql(GET_ALL_WINGS, { name: 'allWings' })(WingList);

