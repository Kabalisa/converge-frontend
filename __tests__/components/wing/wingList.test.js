import React from 'react';
import { shallow } from 'enzyme';

import { WingList } from '../../../src/components/wing/wingList';
import defaultUserRole from '../../../src/fixtures/user';

describe('Tests for SettingOffices', () => {
  const initialProps = {
    allWings: {
      allWings: [{ name: 'epic', id: 1 }],
    },
    user: defaultUserRole,
  };
  const shalloWrapper = shallow(<WingList
    allWings={initialProps.allWings}
    user={defaultUserRole}
  />);

  it('renders the snapshot correctly', () => {
    expect(shalloWrapper).toMatchSnapshot();
  });

  it('should handle WingList component', () => {
    const props = {
      allWings: {
        allWings: [{ name: 'epic', id: 1 }],
      },
      user: defaultUserRole,
    };
    shallow(<WingList allWings={props.allWings} user={defaultUserRole} />);
    expect(props.allWings.allWings).toHaveLength(1);
  });
});
