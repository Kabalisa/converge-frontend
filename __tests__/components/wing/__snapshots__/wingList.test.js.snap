// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for SettingOffices renders the snapshot correctly 1`] = `
<div>
  <div
    className="settings-offices"
  >
    <div
      className="settings-offices-control"
    >
      <MenuTitle
        title="Wings"
      />
      <Apollo(Apollo(AddWing)) />
    </div>
    <div
      className="settings-offices-list"
    >
      <table>
        <ColGroup
          numberOfMiddleColumns={2}
        />
        <TableHead
          titles={
            Array [
              "Wing",
              "Block",
              "Floor",
              "Action",
            ]
          }
        />
        <tbody>
          <DataNotFound />
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
